server/etcdserver/api/membership/downgrade.go

```diff
@@ -64,13 +64,27 @@ func mustDetectDowngrade(lg *zap.Logger, cv *semver.Version, d *DowngradeInfo) {
 	}
 
 	// if the cluster disables downgrade, check local version against determined cluster version.
-	// the validation passes when local version is not less than cluster version
+	// the validation passes
+	// 1. when local version is not less than cluster version
+	// 2. if 1 is not met, when --unsafe-allow-cluster-version-downgrade is enabled
+	//    and local version is 1 minor version less than cluster version
+	//    mutate the cluster version with local version
 	if cv != nil && lv.LessThan(*cv) {
-		lg.Fatal(
-			"invalid downgrade; server version is lower than determined cluster version",
-			zap.String("current-server-version", version.Version),
+		if !allowMinorVersionDowngrade || lv.Major < cv.Major || lv.Minor < cv.Minor-1 {
```

Comment:
Can we still log as warn if we allow downgrades?

```go
downgrade, failDowngrade := false, false
if cv != nil && lv.LessThan(*cv) {
  downgrade, failDowngrade = true, true
}
// if downgrade is enabled, and one minor version down
// safe to not fail (e.g., local version 3.4, cluster version 3.5)
if downgrade && allowMinorVersionDowngrade && lv.Major == cv.Major && && lv.Minor == cv.Minor-1 {
  failDowngrade = false
}
switch {
  case downgrade && failDowngrade:
    lg.Panic("invalid downgrade, not allowed; local server version is lower than determined cluster version",
      zap.String("current-server-version", version.Version),
      zap.String("determined-cluster-version", version.Cluster(cv.String())),
    )
  case downgrade && ! failDowngrade:
    lg.Warn("allowing unsafe downgrade; local server version is lower than determined cluster version",
      zap.String("current-server-version", version.Version),
      zap.String("determined-cluster-version", version.Cluster(cv.String())),
      zap.String("target-cluster-version", version.Cluster(lv.String())),
    )
  case !downgrade:
    // no-op
  default:
    // panic("unknown downgrade check")
}

```

Reply:
much clearer now. will update